File: WeChatFlight/Flight/src/main/java/com/flight/canvas/hero/Bullet1.java
package com.flight.canvas.hero;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;

import com.flight.canvas.state.Counter;
import com.flight.canvas.state.StateConstant.BulletState;
import com.project.wechatflight.R;

public class Bullet1 implements IBullet
{
	private final static int ATK = 1;

	public final static int bulletRes = R.drawable.bullet1;
	
	private Bitmap mBitmap;

	private int mTop, mBottom;
	
	private Rect mRect;

	private BulletState mState;
	
	private Counter mCounter;

	private float positionX, positionY;
	
	public Bullet1(Bitmap bitmap, int bgTop, int bottom)
	{
		this.mBitmap = bitmap;
		this.mTop = bgTop;
		this.mBottom = bottom;
		
		this.mState = BulletState.unStart;
		this.mCounter = new Counter();
		this.positionX = 0;
		this.positionY = 0;
	}

	@Override
	public boolean fireBullet(int x, int y, float durateTime, float frizTime)
	{
		if (BulletState.unStart == mState)
		{
			mCounter.judge = mCounter.caculate("fireBullet", durateTime, frizTime);
			if (mCounter.judge)
			{
				this.mState = BulletState.normal;
				
				this.mRect = new Rect(x - mBitmap.getWidth() / 2, y - mBitmap.getHeight() / 2, x + mBitmap.getWidth() / 2, y + mBitmap.getHeight() / 2);
				this.positionX = mRect.left;
				this.positionY = mRect.top;
				
				return true;
			}
		}
		return false;
	}

	@Override
	public void caculateHeroBullet(float dx, float dy)
	{
		if (BulletState.normal == mState)
		{
			positionX += dx;
			positionY += dy;
			this.mRect.offsetTo((int) positionX, (int) positionY);
			
			if (mRect.bottom < mTop || mRect.top > mBottom)
			{
				this.mState = BulletState.end;
				this.mBitmap = null;
			}
		}
	}

	@Override
	public void drawBullet(Canvas canvas, Paint paint)
	{
		if (BulletState.normal == mState)
		{
			canvas.drawBitmap(mBitmap, mRect.left, mRect.top, paint);
		}
	}
	
	@Override
	public boolean isEnd()
	{
		return BulletState.end == mState ? true : false;
	}
	
	@Override
	public boolean isRunning()
	{
		return BulletState.normal == mState ? true : false;
	}
	
	@Override
	public void disppear()
	{
		this.mState = BulletState.end;
	}

	@Override
	public int getATK()
	{
		return Bullet1.ATK;
	}

	@Override
	public Rect getRect()
	{
		return this.mRect;
	}
}

File: WeChatFlight/Flight/src/main/java/com/flight/canvas/hero/Bullet1.kt
package com.flight.canvas.hero

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import com.flight.canvas.common.InitData
import com.flight.canvas.state.Counter
import com.flight.canvas.state.StateConstant.BulletState
import com.flight.canvas.supply.ISupply
import com.project.wechatflight.R
import java.util.*

class Bullet1(resources: Resources, random: Random, initData: InitData)
    : IBullet(resources, random, initData) {

    override fun clone(resources: Resources, random: Random, initData: InitData): IBullet {
        return Bullet1(resources, random, initData)
    }

    override fun getSourceId(): Int {
        return R.drawable.bullet1
    }


//    override var rect: Rect? = null
//        private set
//    private var mState: BulletState
//    private val mCounter: Counter
//    private var positionX: Float
//    private var positionY: Float
//    override fun fireBullet(x: Int, y: Int, durateTime: Float, frizTime: Float): Boolean {
//        if (BulletState.unStart == mState) {
//            mCounter.judge = mCounter.caculate("fireBullet", durateTime, frizTime)
//            if (mCounter.judge) {
//                mState = BulletState.normal
//                rect = Rect(x - mBitmap!!.width / 2, y - mBitmap!!.height / 2, x + mBitmap!!.width / 2, y + mBitmap!!.height / 2)
//                positionX = rect!!.left.toFloat()
//                positionY = rect!!.top.toFloat()
//                return true
//            }
//        }
//        return false
//    }
//
//    override fun caculateHeroBullet(dx: Float, dy: Float) {
//        if (BulletState.normal == mState) {
//            positionX += dx
//            positionY += dy
//            rect!!.offsetTo(positionX.toInt(), positionY.toInt())
//            if (rect!!.bottom < mTop || rect!!.top > mBottom) {
//                mState = BulletState.end
//                mBitmap = null
//            }
//        }
//    }
//
//    override fun drawBullet(canvas: Canvas, paint: Paint?) {
//        if (BulletState.normal == mState) {
//            canvas.drawBitmap(mBitmap!!, rect!!.left.toFloat(), rect!!.top.toFloat(), paint)
//        }
//    }
//
//    override val isEnd: Boolean
//        get() = if (BulletState.end == mState) true else false
//
//    override val isRunning: Boolean
//        get() = if (BulletState.normal == mState) true else false
//
//    override fun disppear() {
//        mState = BulletState.end
//    }
//
//    companion object {
//        const val aTK = 1
//        const val bulletRes = R.drawable.bullet1
//    }
//
//    init {
//        mState = BulletState.unStart
//        mCounter = Counter()
//        positionX = 0f
//        positionY = 0f
//    }
}